File: sdk/src/agentboom_sdk/cron.py
"""Cron expression parsing and next-fire computation.

One implementation, used by every scheduler in the platform. Before this
existed the reminders mini-app had its own parser and the platform scheduler
had a ``pass  # TODO`` where cron support should have been — which meant a
cron-scheduled job never advanced ``next_fire_at`` and re-fired on every tick.

Format: standard 5-field cron ``minute hour day-of-month month day-of-week``
with ``*``, ``,`` lists, ``a-b`` ranges, ``*/n`` and ``a/n`` steps.
Day-of-week is 0-6 with 0 = Sunday (Unix convention); 7 is also accepted as
Sunday. Both day-of-month and day-of-week restricted means "either matches",
matching Vixie cron.

Usage:
    from agentboom_sdk.cron import next_cron_time, is_valid_cron

    when = next_cron_time("0 9 * * 1-5", after=now, tz_name="Europe/Lisbon")
"""

from __future__ import annotations

import logging
from datetime import datetime, timedelta, timezone
from typing import Dict, List, Optional

log = logging.getLogger("agentboom_sdk.cron")

__all__ = ["parse_cron", "is_valid_cron", "next_cron_time", "CronError"]

# Search horizon. A valid 5-field expression always matches within ~4 years
# (Feb 29 on a leap year is the worst case), so this bound only trips on
# impossible expressions like "0 0 30 2 *".
_MAX_SEARCH_DAYS = 366 * 4

_FIELD_RANGES = (
    ("minute", 0, 59),
    ("hour", 0, 23),
    ("day", 1, 31),
    ("month", 1, 12),
    ("weekday", 0, 6),
)


class CronError(ValueError):
    """Raised when a cron expression cannot be parsed."""


def _parse_field(field: str, name: str, min_val: int, max_val: int) -> List[int]:
    """Parse one cron field into the sorted list of values it matches."""
    values: set[int] = set()
    # Day-of-week accepts 7 = Sunday. Expand in the 0..7 space and wrap
    # afterwards — normalising 7→0 before expansion would turn "5-7" into
    # "5-0" and silently drop Sunday from the range.
    expand_max = 7 if name == "weekday" else max_val

    for part in field.split(","):
        part = part.strip()
        if not part:
            raise CronError(f"{name}: empty element in {field!r}")

        # Split off an optional /step suffix.
        step = 1
        if "/" in part:
            part, _, step_str = part.partition("/")
            try:
                step = int(step_str)
            except ValueError:
                raise CronError(f"{name}: step must be an integer in {field!r}")
            if step <= 0:
                raise CronError(f"{name}: step must be positive in {field!r}")

        if part == "*":
            lo, hi = min_val, expand_max
        elif "-" in part.lstrip("-"):
            lo_str, _, hi_str = part.partition("-")
            try:
                lo, hi = int(lo_str), int(hi_str)
            except ValueError:
                raise CronError(f"{name}: bad range {part!r} in {field!r}")
        else:
            try:
                lo = int(part)
            except ValueError:
                raise CronError(f"{name}: bad value {part!r} in {field!r}")
            # A bare number with a step means "from here to the end of range",
            # e.g. "5/10" in the minute field is 5,15,25,35,45,55.
            hi = expand_max if step > 1 else lo

        if lo < min_val or hi > expand_max or hi < lo:
            raise CronError(
                f"{name}: {part!r} out of range {min_val}-{max_val} in {field!r}"
            )

        values.update(range(lo, hi + 1, step))

    if name == "weekday":
        values = {0 if v == 7 else v for v in values}

    if not values:
        raise CronError(f"{name}: {field!r} matches nothing")
    return sorted(values)


def parse_cron(expr: str) -> Dict[str, List[int]]:
    """Parse a 5-field cron expression into per-field value lists.

    Raises CronError if the expression is malformed.
    """
    if not expr or not expr.strip():
        raise CronError("empty cron expression")

    parts = expr.strip().split()
    if len(parts) != 5:
        raise CronError(
            f"expected 5 fields (minute hour day month weekday), got {len(parts)}: {expr!r}"
        )

    return {
        name: _parse_field(parts[i], name, lo, hi)
        for i, (name, lo, hi) in enumerate(_FIELD_RANGES)
    }


def is_valid_cron(expr: str) -> bool:
    """Return True if the expression parses. Never raises."""
    try:
        parse_cron(expr)
        return True
    except CronError:
        return False


def _resolve_tz(tz_name: str):
    """Return a tzinfo for tz_name, falling back to UTC if unavailable."""
    try:
        from zoneinfo import ZoneInfo

        return ZoneInfo(tz_name)
    except Exception:
        # A typo in SCHEDULER_TZ would otherwise silently reschedule
        # everything in UTC — loud enough to notice, never fatal.
        log.warning("Unknown timezone %r — falling back to UTC", tz_name)
        return timezone.utc


def next_cron_time(
    expr: str,
    after: Optional[datetime] = None,
    tz_name: str = "UTC",
) -> Optional[datetime]:
    """Return the first datetime strictly after ``after`` that matches ``expr``.

    The returned value is timezone-aware and normalised to UTC, so it can be
    written straight into a ``timestamptz`` column. Matching is done in
    ``tz_name`` so "every day at 09:00" means 09:00 local, and follows DST.

    Returns None if the expression is invalid or matches no time within four
    years (e.g. "0 0 30 2 *" — February 30th never happens).
    """
    try:
        fields = parse_cron(expr)
    except CronError:
        return None

    tz = _resolve_tz(tz_name)
    base = after or datetime.now(timezone.utc)
    if base.tzinfo is None:
        base = base.replace(tzinfo=timezone.utc)

    # Start from the next whole minute so we never return `after` itself.
    cursor = base.astimezone(tz).replace(second=0, microsecond=0) + timedelta(minutes=1)

    minutes = fields["minute"]
    hours = fields["hour"]
    days = fields["day"]
    months = fields["month"]
    weekdays = fields["weekday"]

    # Vixie cron semantics: when BOTH day-of-month and day-of-week are
    # restricted, a day matches if EITHER matches. If only one is restricted,
    # only that one applies.
    dom_restricted = len(days) < 31
    dow_restricted = len(weekdays) < 7

    deadline = cursor + timedelta(days=_MAX_SEARCH_DAYS)

    while cursor < deadline:
        if cursor.month not in months:
            # Jump to the 1st of the next month — skips up to 31 days at once.
            cursor = _start_of_next_month(cursor)
            continue

        dom_ok = cursor.day in days
        # Python: Monday=0..Sunday=6. Cron: Sunday=0..Saturday=6.
        dow_ok = ((cursor.weekday() + 1) % 7) in weekdays

        if dom_restricted and dow_restricted:
            day_ok = dom_ok or dow_ok
        elif dom_restricted:
            day_ok = dom_ok
        elif dow_restricted:
            day_ok = dow_ok
        else:
            day_ok = True

        if not day_ok:
            cursor = _start_of_next_day(cursor)
            continue

        if cursor.hour not in hours:
            cursor = _start_of_next_hour(cursor)
            continue

        if cursor.minute in minutes:
            return cursor.astimezone(timezone.utc)

        cursor += timedelta(minutes=1)

    return None


def _start_of_next_day(dt: datetime) -> datetime:
    """Midnight of the following day, DST-safe."""
    naive_next = (dt.replace(tzinfo=None) + timedelta(days=1)).replace(
        hour=0, minute=0, second=0, microsecond=0
    )
    return naive_next.replace(tzinfo=dt.tzinfo)


def _start_of_next_hour(dt: datetime) -> datetime:
    return (dt + timedelta(hours=1)).replace(minute=0, second=0, microsecond=0)


def _start_of_next_month(dt: datetime) -> datetime:
    year, month = (dt.year + 1, 1) if dt.month == 12 else (dt.year, dt.month + 1)
    naive = datetime(year, month, 1)
    return naive.replace(tzinfo=dt.tzinfo)

File: sdk-ts/src/cron.ts
/**
 * 5-field cron parsing + next-fire — port of agentboom_sdk.cron.
 *
 * `minute hour day month weekday`, with `*`, lists, ranges, `*/n` steps.
 * Weekday is 0-6 (0 = Sunday) and 7 is accepted as Sunday, including at
 * the end of ranges ("5-7" keeps Sunday). When both day-of-month and
 * day-of-week are restricted, either matching satisfies (Vixie cron).
 */

export class CronError extends Error {}

const FIELD_RANGES: Array<[string, number, number]> = [
  ["minute", 0, 59],
  ["hour", 0, 23],
  ["day", 1, 31],
  ["month", 1, 12],
  ["weekday", 0, 6],
];

const MAX_SEARCH_DAYS = 366 * 4;

function parseField(field: string, name: string, minVal: number, maxVal: number): number[] {
  const values = new Set<number>();
  // Weekday expands in 0..7 and wraps afterwards — normalising 7→0 before
  // expansion would turn "5-7" into "5-0" and silently drop Sunday.
  const expandMax = name === "weekday" ? 7 : maxVal;

  for (let rawPart of field.split(",")) {
    let part = rawPart.trim();
    if (!part) throw new CronError(`${name}: empty element in ${field}`);
    let step = 1;
    const slash = part.indexOf("/");
    if (slash !== -1) {
      const stepStr = part.slice(slash + 1);
      part = part.slice(0, slash);
      step = Number.parseInt(stepStr, 10);
      if (!Number.isInteger(step) || step <= 0) {
        throw new CronError(`${name}: step must be a positive integer in ${field}`);
      }
    }
    let lo: number;
    let hi: number;
    if (part === "*") {
      lo = minVal;
      hi = expandMax;
    } else if (part.includes("-")) {
      const [loStr, hiStr] = part.split("-");
      lo = Number.parseInt(loStr, 10);
      hi = Number.parseInt(hiStr, 10);
      if (!Number.isInteger(lo) || !Number.isInteger(hi)) {
        throw new CronError(`${name}: bad range ${part} in ${field}`);
      }
    } else {
      lo = Number.parseInt(part, 10);
      if (!Number.isInteger(lo)) throw new CronError(`${name}: bad value ${part} in ${field}`);
      // "5/10" means from here to the end of the range.
      hi = step > 1 ? expandMax : lo;
    }
    if (lo < minVal || hi > expandMax || hi < lo) {
      throw new CronError(`${name}: ${part} out of range ${minVal}-${maxVal} in ${field}`);
    }
    for (let v = lo; v <= hi; v += step) values.add(v);
  }

  const out = [...values].map((v) => (name === "weekday" && v === 7 ? 0 : v));
  if (out.length === 0) throw new CronError(`${name}: ${field} matches nothing`);
  return [...new Set(out)].sort((a, b) => a - b);
}

export interface CronFields {
  minute: number[];
  hour: number[];
  day: number[];
  month: number[];
  weekday: number[];
}

export function parseCron(expr: string): CronFields {
  if (!expr || !expr.trim()) throw new CronError("empty cron expression");
  const parts = expr.trim().split(/\s+/);
  if (parts.length !== 5) {
    throw new CronError(`expected 5 fields (minute hour day month weekday), got ${parts.length}: ${expr}`);
  }
  const fields = {} as CronFields;
  FIELD_RANGES.forEach(([name, lo, hi], i) => {
    fields[name as keyof CronFields] = parseField(parts[i], name, lo, hi);
  });
  return fields;
}

export function isValidCron(expr: string): boolean {
  try {
    parseCron(expr);
    return true;
  } catch {
    return false;
  }
}

function resolveTz(tzName: string): string {
  try {
    new Intl.DateTimeFormat("en-US", { timeZone: tzName });
    return tzName;
  } catch {
    // A typo in the tz would otherwise silently reschedule in UTC.
    console.warn(`[agentboom-sdk] unknown timezone '${tzName}' — falling back to UTC`);
    return "UTC";
  }
}

/** Wall-clock parts of a Date in a timezone. */
function wallParts(date: Date, tz: string) {
  const dtf = new Intl.DateTimeFormat("en-US", {
    timeZone: tz,
    year: "numeric", month: "2-digit", day: "2-digit",
    hour: "2-digit", minute: "2-digit", second: "2-digit",
    hour12: false, weekday: "short",
  });
  const parts: Record<string, string> = {};
  for (const { type, value } of dtf.formatToParts(date)) parts[type] = value;
  const weekdayMap: Record<string, number> = {
    Sun: 0, Mon: 1, Tue: 2, Wed: 3, Thu: 4, Fri: 5, Sat: 6,
  };
  return {
    year: Number.parseInt(parts.year, 10),
    month: Number.parseInt(parts.month, 10),
    day: Number.parseInt(parts.day, 10),
    hour: Number.parseInt(parts.hour, 10) % 24,
    minute: Number.parseInt(parts.minute, 10),
    weekday: weekdayMap[parts.weekday] ?? 0,
  };
}

/** Minutes offset between UTC and the tz at a given instant. */
function tzOffsetMinutes(date: Date, tz: string): number {
  const wall = wallParts(date, tz);
  const asUtc = Date.UTC(wall.year, wall.month - 1, wall.day, wall.hour, wall.minute);
  return Math.round((asUtc - date.getTime()) / 60000);
}

/**
 * First datetime strictly after `after` that matches `expr`, matching in
 * `tz` (DST-aware). Returns a UTC Date, or null when nothing matches
 * within four years.
 */
export function nextCronTime(expr: string, after?: Date, tz = "UTC"): Date | null {
  let fields: CronFields;
  try {
    fields = parseCron(expr);
  } catch {
    return null;
  }
  const zone = resolveTz(tz);
  const base = after ?? new Date();
  // Start from the next whole minute so we never return `after` itself.
  let cursor = new Date(Math.floor(base.getTime() / 60000) * 60000 + 60000);
  const deadline = cursor.getTime() + MAX_SEARCH_DAYS * 86400000;

  while (cursor.getTime() < deadline) {
    const offset = tzOffsetMinutes(cursor, zone);
    const wall = wallParts(cursor, zone);

    if (!fields.month.includes(wall.month)) {
      // Jump to the 1st of the next month (approximate, in the tz).
      const nextMonth = wall.month === 12 ? 1 : wall.month + 1;
      const year = wall.month === 12 ? wall.year + 1 : wall.year;
      const guessUtc = Date.UTC(year, nextMonth - 1, 1, 0, 0) - offset * 60000;
      cursor = new Date(Math.max(guessUtc, cursor.getTime() + 60000));
      continue;
    }

    const domOk = fields.day.includes(wall.day);
    const dowOk = fields.weekday.includes(wall.weekday);
    const domRestricted = fields.day.length < 31;
    const dowRestricted = fields.weekday.length < 7;
    const dayOk = domRestricted && dowRestricted
      ? domOk || dowOk
      : domRestricted ? domOk
      : dowRestricted ? dowOk
      : true;
    if (!dayOk) {
      cursor = new Date(cursor.getTime() + (24 - wall.hour) * 3600000 - wall.minute * 60000);
      continue;
    }
    if (!fields.hour.includes(wall.hour)) {
      cursor = new Date(cursor.getTime() + (60 - wall.minute) * 60000);
      continue;
    }
    if (fields.minute.includes(wall.minute)) return cursor;
    cursor = new Date(cursor.getTime() + 60000);
  }
  return null;
}
